File: log4all/model/group.py
from bson import ObjectId

__author__ = 'Igor Maculan <n3wtron@gmail.com>'


class Group():
    def __init__(self, name, description, permissions=[]):
        self._id = None
        self.name = name
        self.description = description
        self.permissions = permissions

    def __json__(self, request=None):
        json = {'name': self.name, 'description': self.description, 'permissions': self.permissions}
        if self._id is not None:
            json['_id'] = self._id
        return json

    @staticmethod
    def from_bson(bson):
        if bson is None:
            return None
        group = Group(bson['name'], bson.get('description'), bson.get('permissions'))
        group._id = str(bson['_id'])
        return group

    @staticmethod
    def init(db):
        db.groups.ensure_index('_id', unique=True)
        db.groups.ensure_index('name', unique=True)

    def save(self, db):
        json = self.__json__()
        if self._id is not None:
            del json['_id']
            db.groups.update({'_id': ObjectId(self._id)}, json, w=1)
        else:
            id = db.groups.insert(json, w=1)
            if id is not None:
                self._id = str(id)


    @staticmethod
    def get(db, src_query={}):
        return Group.from_bson(db.groups.find_one(src_query))

    @staticmethod
    def delete(db, group_id):
        db.groups.remove(group_id)

    @staticmethod
    def search(db, src_query={}):
        for dbApp in db.groups.find(src_query):
            yield Group.from_bson(dbApp)

File: app/models/group.go
package models

import (
	"gopkg.in/mgo.v2"
	"gopkg.in/mgo.v2/bson"
	//"log"
)

type Group struct {
	BasicModel						`bson:"-"`
	Id				bson.ObjectId	`json:"_id" bson:"_id"`
	Name 			string 			`json:"name"`
	Description 	string			`json:"description"`
	Permissions		[]string		`json:"permissions"`			
}

func CreateGroupIndexes(db *mgo.Database) error{
	index := mgo.Index{
		Key: []string{"name"},
		Unique: true,
		DropDups: true,
		Background: true,
	}
	return db.C("groups").EnsureIndex(index)
}

func (this *Group) Save(db *mgo.Database) error{
	this.Id = bson.NewObjectId()
	return db.C("groups").Insert(this)
}

func (this *Group) Update(db *mgo.Database,id string) error{
	oldGroup,err := GetGroupById(db,id)
	if err != nil {
		return err
	}
	grpUpdErr := db.C("groups").Update(bson.M{"_id":bson.ObjectIdHex(id)},this)
	
	if grpUpdErr !=nil{
		return grpUpdErr
	}
	
	//update users
	users,err := FindUsers(db,bson.M{"groups":bson.M{"$in":[]string{oldGroup.Name}}})
	if err == nil {
		for i:=range(users){
			user := users[i]
			for g:=range(user.Groups){
				uGroup := user.Groups[g]
				if uGroup == oldGroup.Name{
					user.Groups[g] = this.Name
				}
			}
			user.Update(db,user.Id.Hex())
		}
	}
	return err
}

func DeleteGroup(db *mgo.Database, id string) error{
	oldGroup,err := GetGroupById(db,id)
	if err != nil {
		return err
	}
	err = db.C("groups").Remove(bson.M{"_id":bson.ObjectIdHex(id)})
	if err!=nil{
		return err
	}
	users,err := FindUsers(db,bson.M{"groups":bson.M{"$in":[]string{oldGroup.Name}}})
	if err == nil {
		for u:= range(users){
			users[u].Update(db, users[u].Id.Hex())
		}
	}
	return err
}

func GetGroups(db *mgo.Database) ([]Group,error){
	var result []Group
	err := db.C("groups").Find(bson.M{}).All(&result)
	return result,err
}

func GetGroupById(db *mgo.Database, id string) (Group,error){
	var result Group
	err := db.C("groups").Find(bson.M{"_id":bson.ObjectIdHex(id)}).One(&result)
	return result,err
}

func GetGroupByName(db *mgo.Database, name string) (Group,error){
	var result Group
	err := db.C("groups").Find(bson.M{"name":name}).One(&result)
	return result,err
}

